docs: add zkVM technical specification (#1067)
risc0/risc0: website/docs/zkvm/developer-guide/zkvm-specification.md

# zkVM Technical Specification

> Like many other execution environments, the zkVM it has unique characteristics

Reviewer: ```suggestion
Like many other execution environments, the zkVM has unique characteristics
```

unaligned addresses. So all addresses must be multiples of `0x4` and all code
must be mapped to addresses divisible by `0x4`.

We extend the `RV32IM` ISA by using the `ECALL` instruction. We do this in order
to add instructions that are specific to zero knowledge computing.

The zkVM does not support interrupts and there is no built-in notion of a
scheduler. The zkVM runs programs using a single-thread environment without
preemption, so there is also no support for atomic instructions. If the
execution raises an exception such as an unaligned access, the execution
terminates without executing exception handlers.

## zkVM Memory Layout

The following table summarizes the layout of the zkVM memory

| Address Range             | Size   | Name          | Description                                                                                   |
| ------------------------- | ------ | ------------- | --------------------------------------------------------------------------------------------- |
| `0x00000000 - 0x000003ff` | 1 KB   | Invalid page  | This page of memory is invalid, so that dereferencing a NULL address will result in a failure |

> | `0x00000400 - 0x0BFFFFFF` | 192 MB | Guest Memory  | Contains code, heap, and stack of the guest program                                           |

Reviewer: ```suggestion
| `0x00000400 - 0x0BFFFFFF` | 192 MB | User Memory  | Contains code, heap, and stack of the guest program                                           |
```

| `0x0C000000 - 0x0CFFFFFF` | 16 MB  | System Memory | Contains RISC-V registers and non-leaf Merkle tree nodes for the page table                   |
| `0x0D000000 - 0x0DFFFFFF` | 16 MB  | Page Table    | Contains the Page Table                                                                       |

## zkVM Memory Model

> The zkVM executes instruction in the order they appear in the ELF. In other

Reviewer: ```suggestion
The zkVM executes instructions in order; in other
```
